sherpa/utils/random.py

```diff
@@ -0,0 +1,290 @@
+#
+#  Copyright (C) 2023
+#  Smithsonian Astrophysical Observatory
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+"""
+Basic routines for handling random numbers.
+
+A number of routines have been added to support writing code that
+allows both the legacy (pre 1.17) NumPy random API to be used
+(with rng set to None) or the new generator approach.
+
+"""
+
+import numpy as np
+
+# Copy of sherpa.utils.SherpaFloat
+#
+SherpaFloat = np.float_
+
+
+def poisson_noise(x, rng=None):
+    """Draw samples from a Poisson distribution.
+
+
+    Parameters
+    ----------
+    x : scalar or array
+       The expectation value for the distribution.
+    rng : np.random.Generator, np.random.RandomState, or None, optional
+        If set, the generator is used to create the random numbers. If
+        not set then the legacy numpy RandomState instance is used.
+
+    Returns
+    -------
+    out : scalar or array
+       A random realisation of the input array, drawn from
+       the Poisson distribution, as a `SherpaFloat`.
+
+    Notes
+    -----
+    The distribution is calculated by the NumPy poisson generator
+    and all values zero or less are replaced by zero.
+
+    Examples
+    --------
+
+    >>> poisson_noise([10, 20, 5])
+    array([ 13.,  21.,   6.])
+
+    >>> rng = np.random.default_rng(17389)
+    >>> poisson_noise([1, 2, 2, 3], rng=rng)
+    array([2., 2., 5., 3.])
+
+    """
+
+    if rng is None:
+        simulate = np.random.poisson
+    else:
+        simulate = rng.poisson
+
+    x = np.asarray(x)
+    if x.shape == ():
+        x = SherpaFloat(x)
+        if x <= 0.0:
+            x = 0.0
+        else:
+            x = x = simulate(x)
+
+        return SherpaFloat(x)
+
+    x_out = np.zeros(x.shape, SherpaFloat)
+    good = np.where(x > 0.0)
+    x_out[good] = simulate(x[good])
+    return x_out
+
+
+def random(rng):
+    """Create a random value [0, 1.0)
+
+    Parameters
+    ----------
+    rng : np.random.Generator, np.random.RandomState, or None, optional
+        If set, the generator is used to create the random numbers. If
+        not set then the legacy numpy RandomState instance is used.
```

Comment:
I struggled to explain some of this when writing, so thanks for looking Will rework.

Reply:
My problem is that "If not set then the legacy numpy RandomState" text is technically incorrect (this may be an issue you raise elsewhere) - what is actually used are the numpy.random.xxx routines, so maybe just "legacy numpy random routine" would be better (and then should it label the actual routine, which we can do in this case ?)

Reply:
Hopefully the new text is better.